File: samples/java/DeliveryDispatch/Server/CourierSpotNode/src/main/java/systems/zlink/samples/deliverydispatch/server/courierspotnode/spots/handlers/BindCourierSessionActorHandler.java
package systems.zlink.samples.deliverydispatch.server.courierspotnode.spots.handlers;

import systems.zlink.framework.ZLinkMessageContext;
import systems.zlink.framework.spots.ZLinkEntrySpotActorRequestHandler;
import systems.zlink.samples.deliverydispatch.server.courierspotnode.CourierActor;
import systems.zlink.samples.deliverydispatch.server.courierspotnode.spots.CourierEntrySpot;
import systems.zlink.samples.deliverydispatch.shared.contracts.Messages;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

public final class BindCourierSessionActorHandler
        implements ZLinkEntrySpotActorRequestHandler<
                CourierEntrySpot,
                CourierActor,
                Messages.BindCourierSessionReq,
                Messages.BindCourierSessionRes> {
    @Override
    public CompletionStage<Messages.BindCourierSessionRes> handle(
            CourierEntrySpot entrySpot,
            CourierActor actor,
            ZLinkMessageContext context,
            Messages.BindCourierSessionReq request) {
        System.out.println("deliverydispatch-courier bind-relayed courier=" + request.courierId());
        return CompletableFuture.completedFuture(
                new Messages.BindCourierSessionRes(
                        request.courierId(), request.actor(), request.sessionRoute()));
    }
}

File: samples/kotlin/DeliveryDispatch/Server/CourierSpotNode/src/main/kotlin/systems/zlink/samples/kotlin/deliverydispatch/server/courierspotnode/spots/handlers/BindCourierSessionActorHandler.kt
package systems.zlink.samples.kotlin.deliverydispatch.server.courierspotnode.spots.handlers

import systems.zlink.framework.ZLinkMessageContext
import systems.zlink.framework.kotlin.ZLinkSuspendingEntrySpotActorRequestHandler
import systems.zlink.samples.kotlin.deliverydispatch.server.courierspotnode.CourierActor
import systems.zlink.samples.kotlin.deliverydispatch.server.courierspotnode.spots.CourierEntrySpot
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.BindCourierSessionReq
import systems.zlink.samples.kotlin.deliverydispatch.shared.contracts.BindCourierSessionRes

class BindCourierSessionActorHandler :
    ZLinkSuspendingEntrySpotActorRequestHandler<
        CourierEntrySpot,
        CourierActor,
        BindCourierSessionReq,
        BindCourierSessionRes,
    > {
    override suspend fun handle(
        entrySpot: CourierEntrySpot,
        actor: CourierActor,
        context: ZLinkMessageContext,
        request: BindCourierSessionReq,
    ): BindCourierSessionRes {
        val response =
            BindCourierSessionRes(
                courierId = request.courierId,
                actor = requireNotNull(request.actor) { "BindCourierSessionReq.actor is required" },
                sessionRoute =
                    requireNotNull(request.sessionRoute) {
                        "BindCourierSessionReq.sessionRoute is required"
                    },
            )
        println("deliverydispatch-courier bind-relayed courier=${request.courierId}")
        return response
    }
}
